### PR TITLE
Friedman Rafsky PR

### PR DESCRIPTION
**Pull request by zdbzdb123123**
@sampan501 

closes #239

### CODE REVIEW

**Comment by sampan501** [comment]
I'll review after you fix the build issues

**Comment by sampan501** [comment]
Make sure you run black formatting on your files, that is the only problem regarding building

**Comment by zdbzdb123123** [comment]
The tutorial folder in hyppo includes one large tutorial file for all of the independence tests. Do you want me to copy what I've written into that file or add it as a separate tutorial file? As far as the black reformatting I just installed black and ran black friedmanRafsky.py from terminal but if that is not what you're looking for just let me know

**Comment by zdbzdb123123** [comment]
When I run black friedmanRafsky.py --diff, I'm being told the file would remain unchanged. However it looks like for some reason the check is failing when run as part of the PR check

**Comment by sampan501** [comment]
Make sure that `black` is fully up to date

**Comment by sampan501** [comment]
> The tutorial folder in hyppo includes one large tutorial file for all of the independence tests. Do you want me to copy what I've written into that file or add it as a separate tutorial file?

Yes copy what you have to the file

**Comment by zdbzdb123123** [comment]
Looks like black wasn't updated, file was reformatted and I'll copy over the tutorial into the larger file. Thanks

File: hyppo/independence/friedmanRafsky.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
+import random
+from numba import jit
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+class FriedmanRafsky(IndependenceTest):
+
+    r"""
+    Friedman-Rafksy (FR) test statistic and p-value.
+    This is a multivariate extension of the Wald-Wolfowitz
+    runs test for randomness. The normal concept of a 'run'
+    is replaced by a minimum spanning tree calculated between
+    the points in respective data sets with edge weights defined
+    as the Euclidean distance between two such points. After MST
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: Define abbreviation before using it

File: hyppo/independence/friedmanRafsky.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    has been determined, all edges such that both corresponding
+    nodes do not belong to the same class are severed and the
+    number of independent resulting trees is counted. This test is
+    consistent against similar tests
+    :footcite:p:`GSARMultivariateAdaptationofWaldWolfowitzTest`
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: this is in `refs.bib` right?

File: hyppo/independence/friedmanRafsky.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    """
+
+    def __init__(self, **kwargs):
+
+        IndependenceTest.__init__(self, **kwargs)
+
+    class Graph:
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: why is this a class in a class? This should be able to exist as a standalone class at the bottom of the file right?

File: hyppo/independence/friedmanRafsky.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+        r"""
+        Helper class that finds the MST for a given dataset using
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: Formatting off, see: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

File: hyppo/independence/friedmanRafsky.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        Kruskal's MST algorithm. Takes in node lables and edge weights
+        before returning all pairs of connected node labels in the
+        resultant MST.
+        Parameters
+        ----------
+        i,j,weight : int, int, float
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: `i`, `j`, and `weight` are not parameters for this class. Also where is `vertex`? What is `self2`?

File: hyppo/independence/friedmanRafsky.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+            Input node indeces i, j, and corresponding Euclidean distance
+            edge weight between them.
+        Returns
+        -------
+        result : list
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: I don't believe this class returns anything

File: hyppo/independence/friedmanRafsky.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+        """
+
+        def __init__(self2, vertex):
+
+            self2.V = vertex
+            self2.graph = (
+                []
+            )  # Empty matrix for holding vertices and weights connecting them
+
+        def add_edge(self2, v1, v2, w):
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: add docstring for this method, at the very least what it does

File: hyppo/independence/friedmanRafsky.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+
+            self2.graph.append(
+                [v1, v2, w]
+            )  # Add method for creating edges between vertices
+
+        def search(
+            self2, parent, i
+        ):  # Method for determining location of vertex in existing tree
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: comment should be a docstring

File: hyppo/independence/friedmanRafsky.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+            if parent[i] == i:
+                return i
+
+            return self2.search(parent, parent[i])
+
+        def apply_union(
+            self2, parent, rank, x, y
+        ):  # Method for deleting and merging branches
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: comment should be a docstring

File: hyppo/independence/friedmanRafsky.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+        def add_edge(self2, v1, v2, w):
+
+            self2.graph.append(
+                [v1, v2, w]
+            )  # Add method for creating edges between vertices
+
+        def search(
+            self2, parent, i
+        ):  # Method for determining location of vertex in existing tree
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+        def __init__(self2, vertex):
+
+            self2.V = vertex
+            self2.graph = (
+                []
+            )  # Empty matrix for holding vertices and weights connecting them
+
+        def add_edge(self2, v1, v2, w):
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+            Input node indeces i, j, and corresponding Euclidean distance
+            edge weight between them.
+        Returns
+        -------
+        result : list
+            List of pairs of nodes connected in final MST.
+        """
+
+        def __init__(self2, vertex):
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+            if rank[xroot] < rank[yroot]:
+                parent[xroot] = yroot
+            elif rank[xroot] > rank[yroot]:
+                parent[yroot] = xroot
+            else:
+                parent[yroot] = xroot
+                rank[xroot] += 1
+
+        def kruskal(self2):
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+            while e < self2.V - 1:
+                v1, v2, w = self2.graph[i]
+                i = i + 1
+                x = self2.search(parent, v1)
+                y = self2.search(parent, v2)
+                if x != y:
+                    e = e + 1
+                    result.append([v1, v2])
+                    self2.apply_union(parent, rank, x, y)
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+            return result
+
+    def pval(self, perm_stat, true_stat):
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+        r"""
+        Helper function that calculates the Friedman Rafksy p-value.
+        """
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: remove space

File: hyppo/independence/friedmanRafsky.py (hunk 15 above)
Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: add docstring

File: hyppo/independence/friedmanRafsky.py [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+            else:
+                parent[yroot] = xroot
+                rank[xroot] += 1
+
+        def kruskal(self2):
+
+            result = []
+            i, e = 0, 0
+            self2.graph = sorted(self2.graph, key=lambda item: item[2])
+            parent = []
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: can probably compress this using NumPy arrays

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: 88-92

File: hyppo/independence/friedmanRafsky.py [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+            rank = []
+            for node in range(self2.V):
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: can you send the code where this is ported from? Seems there might be a more efficient way of doing things

File: hyppo/independence/friedmanRafsky.py [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+        """
+
+        pvalue = (np.sum(perm_stat <= true_stat) + 1) / (len(perm_stat) + 1)
+
+        return pvalue
+
+    def test(self, data, labels, perm, algorithm):
+
+        r"""
+        Function to take in data, labels, nperm and return both test-statistic and
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: formatting is off for this docstring again. See above comment

File: hyppo/independence/friedmanRafsky.py [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+
+            return result
+
+    def pval(self, perm_stat, true_stat):
+
+        r"""
+        Helper function that calculates the Friedman Rafksy p-value.
+        """
+
+        pvalue = (np.sum(perm_stat <= true_stat) + 1) / (len(perm_stat) + 1)
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: what is this doing that the existing `perm_test` function in `hyppo` is not?

File: tutorials/independence.py [hunk 23]
```diff
@@ -313,3 +313,134 @@
 # These tests runs like :ref:`any other test<general indep>`.
 #
 # .. _[1]: https://link.springer.com/article/10.1007/s10182-020-00378-1
+
+r"""
```
Comment (sampan501) [inline tutorials/independence.py]: This tutorial is not showing up on the website. Make sure that you are formatting it correctly. See other sections in the independence tutorial for help with this

File: hyppo/independence/friedmanRafsky.py [hunk 24]
```diff
@@ -0,0 +1,297 @@
...
+            j += 1
+
+        MST_connections = prim(G)
+
+    return MST_connections
+
+
+@jit(nopython=True, cache=True)
+def num_runs(labels, MST_connections):
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: extra space

File: hyppo/independence/friedmanRafsky.py [hunk 25]
```diff
@@ -0,0 +1,297 @@
...
+    r"""
+    Helper function to determine number of independent
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: formatting is off for the docstring again

File: hyppo/independence/friedmanRafsky.py [hunk 26]
```diff
@@ -0,0 +1,297 @@
...
+
+    for x in MST_connections:
+        if labels[x[0]] != labels[x[1]]:
+            run_count += 1
+
+    return run_count
+
+
+@jit(nopython=True, cache=True)
+def permutation(perm, labels, MST_connections):
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: what is this doing that `perm_test` does not?

File: hyppo/independence/friedmanRafsky.py [hunk 27]
```diff
@@ -0,0 +1,297 @@
...
+        MST_connections = g.kruskal()
+
+    if algorithm == "Prim":
+
+        G = zeros((len(data[0]), len(data[0])))
+
+        for i in range(len(data[0])):
+            j = i + 1
+
+        while j <= (len(data[0]) - 1):
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: this would be more clear as a for loop

File: hyppo/independence/friedmanRafsky.py [hunk 28]
```diff
@@ -0,0 +1,297 @@
...
+        MST_connections.append([x, y])
+        selected[y] = True
+        no_edge += 1
+
+    return MST_connections
+
+
+@jit(nopython=True, cache=True)
+def MST(data, algorithm):
+
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: extra space

File: hyppo/independence/friedmanRafsky.py [hunk 29]
```diff
@@ -0,0 +1,297 @@
...
+    r"""
+    Helper function to read input data and calculate Euclidean distance
```
Comment (sampan501) [inline hyppo/independence/friedmanRafsky.py]: spacing is off on this docstring. Also, formatting is off

**Review by sampan501 — CHANGES_REQUESTED** [review]
This class is not added to the API reference part of the docs on the website. Look through the `docs/` folder to see how to do that

Where are the unit tests for this code?

There are bunch of changes that need to be made. Let me know when you are done and I'll take a second pass at it

File: hyppo/independence/friedmanRafsky.py (hunk 3 above)
Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: How would I reference a separate R package (seems like everything else is referencing articles)

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: Well all I was asking is whether or not the reference you have here is in the bibtex file. As for citing R package, I would just look up how to cite packages in bibtex

File: hyppo/independence/friedmanRafsky.py (hunk 20 above)
Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: class Graph:
    def __init__(self, vertex):
        self.V = vertex
        self.graph = []
 
    def add_edge(self, u, v, w):
        self.graph.append([u, v, w])
 
 
    def search(self, parent, i):
        if parent[i] == i:
            return i
        return self.search(parent, parent[i])
 
    def apply_union(self, parent, rank, x, y):
        xroot = self.search(parent, x)
        yroot = self.search(parent, y)
        if rank[xroot] < rank[yroot]:
            parent[xroot] = yroot
        elif rank[xroot] > rank[yroot]:
            parent[yroot] = xroot
        else:
            parent[yroot] = xroot
            rank[xroot] += 1
 
  
    def kruskal(self):
        result = []
        i, e = 0, 0
        self.graph = sorted(self.graph, key=lambda item: item[2])
        parent = []
        rank = []
        for node in range(self.V):
            parent.append(node)
            rank.append(0)
        while e < self.V - 1:
            u, v, w = self.graph[i]
            i = i + 1
            x = self.search(parent, u)
            y = self.search(parent, v)
            if x != y:
                e = e + 1
                result.append([u, v, w])
                self.apply_union(parent, rank, x, y)
        for u, v, weight in result:
            print("Edge:",u, v,end =" ")
            print("-",weight)

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: I meant a link. That is the same that is in your PR

Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: My bad, https://www.algotree.org/algorithms/minimum_spanning_tree/kruskals/

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: Ok, make sure you heavily comment the code and move the graph class outside the test class

Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: Will do, finishing up some other edits before re-committing

File: hyppo/independence/friedmanRafsky.py (hunk 26 above)
Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: It was my understanding that perm_test rearranged the y values corresponding to given x values while all labels remained constant (i.e. a permuted x value still has an x label, y value still has a y label). This instead permutes the labels and then determines the new run count from the MST based on the same labels to calc the test statistic. If I'm misinterpreting the usage here just let me know

File: tutorials/independence.py (hunk 23 above)
Comment (zdbzdb123123) [reply tutorials/independence.py]: Tried reformatting, let me know if issue appears fixed

File: hyppo/independence/friedmanRafsky.py (hunk 26 above)
Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: What do you mean by labels?

For independence tests, `perm_test` computes the p-values by permuting the values of matrix `y` (which can contain labels) and recomputing the test statistic.

Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: The labels would be some value (0) for x points and some value (1) for y points. I guess I was misinterpreting the perm_test code. How would I incorporate the existing perm_test function into the code

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: what you described is equivalent to the `perm_test` code minus the labels matrix. Permuting the labels like this is equivalent to permuting `y`.

Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: perm_test takes in data x such that each row is a data point it looks like while I've written the code to operate as if each column was a data point. Do you want me to rewrite the test itself to work for some x and y where rows of x are the data points of the combined samples and y is a column vector of the labels?

Comment (zdbzdb123123) [reply hyppo/independence/friedmanRafsky.py]: Also to compute the test statistic for the true case I need to permute the labels to compare to the random run counts so I think I would have to permute anyways

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: > perm_test takes in data x such that each row is a data point it looks like while I've written the code to operate as if each column was a data point. Do you want me to rewrite the test itself to work for some x and y where rows of x are the data points of the combined samples and y is a column vector of the labels?

just transpose the inputs

Comment (sampan501) [reply hyppo/independence/friedmanRafsky.py]: > Also to compute the test statistic for the true case I need to permute the labels to compare to the random run counts so I think I would have to permute anyways

you are. What I am saying is that both ways are equivalent and valid ways to permute a dataset

**Comment by zdbzdb123123** [comment]
Just updated the code for perm test and reformatted/changed variable names to fit better with other tests in hyppo. I was getting caught up on needing to standardize the run count for the test statistic but realized it's always demeaned and standardized so the relationship would still stand across the board. Looks like the formatting flopped so I'll black reformat and re-upload

**Comment by zdbzdb123123** [comment]
Just black formatted and am going to get the unit tests done. Let me know if anything still looks off about the test itself

**Comment by zdbzdb123123** [comment]
Not entirely sure why it is no longer building if you wouldn't mind taking a look

**Comment by sampan501** [comment]
> Not entirely sure why it is no longer building if you wouldn't mind taking a look

Well, the CircleCI logs say that the unit tests for your code are failing. [This](https://app.circleci.com/pipelines/github/neurodata/hyppo/767/workflows/127d3c3f-3458-45a4-b287-6d6e8e4a738f/jobs/3985/tests) tells you why. Looks to me that you didn't add your class [here](https://github.com/zdbzdb123123/hyppo/blob/main/hyppo/independence/__init__.py), or to the dictionary in that file. Also, you are calling the module name of `friedmanRafsky` instead of your class object `FriedmanRafsky`.

Also, side note, make sure you use python conventions and package conventions for naming files. File should be called `friedman_rafsky.py` and the test file should be `test_friendman_rafsky.py`

**Comment by zdbzdb123123** [comment]
Looks like the issue now is numba can't accept lists of lists as inputs to functions. Should I move the function inside of the greater class (removing it from the jit wrapper) or rewrite the code such that it avoids a list of lists.

**Comment by sampan501** [comment]
yes

**Comment by zdbzdb123123** [comment]
Feel free to take another run through whenever. I just removed Kruskal's algorithm from the code (potential later addition) since the checks didn't seem to like the additional class and Prim's had a faster asymptotic runtime anyways for the same result (also more efficient in general). I wrote unit tests to reproduce results from the original r-package in addition to confirming that results were constant across diff runnings of the test. The codecov is saying tests don't cover the 2 jit functions (prim and MST) even though both are used to calculate the test statistic so if you have any thoughts on how to get around that, would be much appreciated. Thanks

File: hyppo/independence/friedman_rafsky.py [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+    Friedman-Rafksy (FR) test statistic and p-value.
+    This is a multivariate extension of the Wald-Wolfowitz
+    runs test for randomness. The normal concept of a 'run'
+    is replaced by a minimum spanning tree (MST) calculated between
+    the points in respective data sets with edge weights defined
+    as the Euclidean distance between two such points. After MST
+    has been determined, all edges such that both corresponding
+    nodes do not belong to the same class are severed and the
+    number of independent resulting trees is counted. This test is
+    consistent against similar tests.
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:20]: Add notes section discussing how to calculate the test statistic, and a references section citing the correct paper. As [here](https://github.com/neurodata/hyppo/blob/f7a1f6c3ba214884c7f508ee60159b99962a3e9c/hyppo/independence/rv.py#L19)

File: hyppo/independence/friedman_rafsky.py [hunk 31]
```diff
@@ -0,0 +1,206 @@
...
+
+    def __init__(self, **kwargs):
+
+        IndependenceTest.__init__(self, **kwargs)
+
+    def num_runs(self, labels, MST_connections):
+        r"""
+        Helper function to determine number of independent
+        'runs' from MST connections.
+        Parameters
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:49]: add newline above 31, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 32]
```diff
@@ -0,0 +1,206 @@
...
+        ----------
+        labels : ndarry
+            Lables corresponding to respective classes of samples.
+        MST_connections: list
+            List containing pairs of points connected in final MST.
+        Returns
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:56]: add newline above 37, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 33]
```diff
@@ -0,0 +1,206 @@
...
+        for x in MST_connections:
+            if labels[x[0]] != labels[x[1]]:
+                run_count += 1
+
+        return run_count
+
+    def statistic(self, x, y):
+        r"""
+        Helper function that calculates the Friedman Rafksy test statistic.
+        Parameters
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:74]: add newline above 54, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 34]
```diff
@@ -0,0 +1,206 @@
...
+        ----------
+        x,y : ndarray
+            Input data matrices. ``x`` and ``y`` must have the same number of
+            rows. That is, the shapes must be ``(n, p)`` and ``(n, 1)`` where
+            `n` is the number of combined samples and `p` is the number of
+            dimensions. ``y`` is the array of labels corresponding to the two
+            samples, respectively.
+        Returns
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:83]: add newline above 62, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 35]
```diff
@@ -0,0 +1,206 @@
...
+                run_count += 1
+
+        return run_count
+
+    def statistic(self, x, y):
+        r"""
+        Helper function that calculates the Friedman Rafksy test statistic.
+        Parameters
+        ----------
+        x,y : ndarray
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: `ndarray of float`

File: hyppo/independence/friedman_rafsky.py [hunk 36]
```diff
@@ -0,0 +1,206 @@
...
+
+        IndependenceTest.__init__(self, **kwargs)
+
+    def num_runs(self, labels, MST_connections):
+        r"""
+        Helper function to determine number of independent
+        'runs' from MST connections.
+        Parameters
+        ----------
+        labels : ndarry
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: ndarray of float

File: hyppo/independence/friedman_rafsky.py [hunk 37]
```diff
@@ -0,0 +1,206 @@
...
+        y,
+        reps=1000,
+        workers=1,
+        is_distsim=False,
+        perm_blocks=None,
+        random_state=None,
+    ):
+        r"""
+        Calculates the Friedman Rafsky test statistic and p-value.
+        Parameters
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:107]: add newline above 87, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 38]
```diff
@@ -0,0 +1,206 @@
...
+            If None, all samples can be permuted with one another. Requires `n`
+            rows. At each column, samples with matching column value are
+            recursively partitioned into blocks of samples. Within each final
+            block, samples are exchangeable. Blocks of samples from the same
+            partition are also exchangeable between one another. If a column
+            value is negative, that block is fixed and cannot be exchanged.
+        random_state : int, default: None
+            The random_state for permutation testing to be fixed for
+            reproducibility.
+        Returns
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:125]: add newline above 112, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 39]
```diff
@@ -0,0 +1,206 @@
...
+        -------
+        stat : float
+            The computed Friedman Rafsky statistic.
+        pvalue : float
+            The computed Friedman Rafsky p-value.
+        null_dist : array
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: you don't return this, why is this here?

File: hyppo/independence/friedman_rafsky.py [hunk 40]
```diff
@@ -0,0 +1,206 @@
...
+    nodes do not belong to the same class are severed and the
+    number of independent resulting trees is counted. This test is
+    consistent against similar tests.
+    """
+
+    def __init__(self, **kwargs):
+
+        IndependenceTest.__init__(self, **kwargs)
+
+    def num_runs(self, labels, MST_connections):
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: add underscore before

File: hyppo/independence/friedman_rafsky.py [hunk 41]
```diff
@@ -0,0 +1,206 @@
...
+        null_dist : array
+            The null distribution of Friedman Rafsky test statistics.
+        """
+
+        stat, pvalue, null_dist = perm_test(
+            self.statistic, x, y, reps, workers, is_distsim, perm_blocks, random_state
+        )
+        x = np.transpose(x)
+        labels = np.transpose(y)
+        MST_connections = MST(x, labels)
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: I'm confused what you are doing in 125-129. `perm_test` already returns the observed statistic

File: hyppo/independence/friedman_rafsky.py [hunk 42]
```diff
@@ -0,0 +1,206 @@
...
+
+        return IndependenceTestOutput(stat, pvalue)
+
+
+@jit(nopython=True, cache=True)
+def prim(weight_mat, labels):
+    r"""
+    Helper function to read weighted matrix input and compute minimum
+    spanning tree via Prim's algorithm.
+    Parameters
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:154]: add newline above 140, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 43]
```diff
@@ -0,0 +1,206 @@
...
+    ----------
+    weight_mat : ndarry
+        Weighted connection matrix.
+    labels : ndarry
+        Lables corresponding to respective classes of samples.
+    Returns
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:161]: add newline above 146, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 44]
```diff
@@ -0,0 +1,206 @@
...
+
+    return MST_connections
+
+
+@jit(nopython=True, cache=True)
+def MST(x, labels):
+    r"""
+    Helper function to read input data and calculate Euclidean distance
+    between each possible pair of points before finding MST.
+    Parameters
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:199]: add newline above 183, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 45]
```diff
@@ -0,0 +1,206 @@
...
+    ----------
+    data : ndarry
+        Dataset such that each column corresponds to a point of data.
+    labels : ndarry
+        Lables corresponding to respective classes of samples.
+    Returns
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:206]: add newline above 189, otherwise sphinx won't render properly

File: hyppo/independence/friedman_rafsky.py [hunk 46]
```diff
@@ -0,0 +1,206 @@
...
+
+
+@jit(nopython=True, cache=True)
+def MST(x, labels):
+    r"""
+    Helper function to read input data and calculate Euclidean distance
+    between each possible pair of points before finding MST.
+    Parameters
+    ----------
+    data : ndarry
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: you called it `x`. It is `ndarray of float`

File: hyppo/independence/friedman_rafsky.py [hunk 47]
```diff
@@ -0,0 +1,206 @@
...
+        Dataset such that each column corresponds to a point of data.
+    labels : ndarry
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: `ndarray of int`

File: hyppo/independence/friedman_rafsky.py [hunk 48]
```diff
@@ -0,0 +1,206 @@
...
+        Lables corresponding to respective classes of samples.
+    Returns
+    -------
+    MST_connections : list
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py:208]: `list of float`

File: hyppo/independence/friedman_rafsky.py [hunk 49]
```diff
@@ -0,0 +1,206 @@
...
+
+
+@jit(nopython=True, cache=True)
+def prim(weight_mat, labels):
+    r"""
+    Helper function to read weighted matrix input and compute minimum
+    spanning tree via Prim's algorithm.
+    Parameters
+    ----------
+    weight_mat : ndarry
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: `ndarray of float`

File: hyppo/independence/friedman_rafsky.py [hunk 50]
```diff
@@ -0,0 +1,206 @@
...
+        Weighted connection matrix.
+    labels : ndarry
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: `ndarray of int`

File: hyppo/independence/friedman_rafsky.py [hunk 51]
```diff
@@ -0,0 +1,206 @@
...
+        Lables corresponding to respective classes of samples.
+    Returns
+    -------
+    MST_connections : list
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: `list of float`

File: hyppo/independence/friedman_rafsky.py [hunk 52]
```diff
@@ -0,0 +1,206 @@
...
+        return stat
+
+    def test(
+        self,
+        x,
+        y,
+        reps=1000,
+        workers=1,
+        is_distsim=False,
+        perm_blocks=None,
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: remove this

File: hyppo/independence/friedman_rafsky.py [hunk 53]
```diff
@@ -0,0 +1,206 @@
...
+
+        return stat
+
+    def test(
+        self,
+        x,
+        y,
+        reps=1000,
+        workers=1,
+        is_distsim=False,
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: remove this

File: hyppo/independence/friedman_rafsky.py [hunk 54]
```diff
@@ -0,0 +1,206 @@
...
+            `n` is the number of combined samples and `p` is the number of
+            dimensions. ``y`` is the array of labels corresponding to the two
+            samples, respectively.
+        reps : int, default: 1000
+            The number of replications used to estimate the null distribution
+            when using the permutation test used to calculate the p-value.
+        workers : int, default: 1
+            The number of cores to parallelize the p-value computation over.
+            Supply ``-1`` to use all cores available to the Process.
+        perm_blocks : None or ndarray, default: None
```
Comment (sampan501) [inline hyppo/independence/friedman_rafsky.py]: remove this

File: tutorials/independence.py [hunk 55]
```diff
@@ -313,4 +313,151 @@
 # information.
 # These tests runs like :ref:`any other test<general indep>`.
 #
+########################################################################################
+#Friedman Rafsky Test for Randomness
```
Comment (sampan501) [inline tutorials/independence.py]: add space after `#` on each line, or your tutorial won't show up on the website

File: tutorials/independence.py [hunk 56]
```diff
@@ -313,4 +313,151 @@
...
+#to that of the euclidean distance between the two points. The number of runs is then determined by severing 
+#each edge of the MST for which the points connected do not belong to the same family. Similarly to the univariate 
+#case, the labels associated with each node are then permuted randomly a large number of times, we can 
+#compare our number of true runs to the random distribution of runs to determine the multivariate test 
+#statistic and p-value. To further understand, let us consider an exmaple in the 2-dimensional case. 
+#
+#
+########################################################################################
+#
+from IPython.display import Image
```
Comment (sampan501) [inline tutorials/independence.py]: use this instead: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/images.html

**Review by sampan501 — CHANGES_REQUESTED** [review]
`jit` files aren't able to be checked by codecov, so add as ` # pragma: no cover
` I do [here](https://github.com/neurodata/hyppo/blob/f7a1f6c3ba214884c7f508ee60159b99962a3e9c/hyppo/independence/dcorr.py#L415)

**Comment by zdbzdb123123** [comment]
Made all requested edits, I assume I'll have to fix some new build issues but as soon as I resolve those I will add the citations and let you know for a final run through

**Comment by zdbzdb123123** [comment]
Alright so it looks like the only issue now is that, by removing is_distsim from the arguments for the test function, it's defaulting to true. But when I try to manually set it to false inside of the function, it's returning an error that perm_test is getting multiple arguments for is_distsim? Additionally, when I black reformat the file it removes all spacing added between parameters and returns statements in the documentation. Was wondering if you had a workaround for that issue

**Comment by zdbzdb123123** [comment]
Never mind the is_distsim was an easy fix, just let me know what you want me to do about black removing the spaces and we should be good to go

**Review by sampan501 — CHANGES_REQUESTED** [review]
You have a few doc rendering issues according to this: https://deploy-preview-239--hyppo.netlify.app/api/generated/hyppo.independence.friedmanrafsky#hyppo.independence.FriedmanRafsky. Once you fix those, we should be good to merge

**Review by sampan501 — CHANGES_REQUESTED** [review]
Still some errors here: https://deploy-preview-239--hyppo.netlify.app/tutorials/independence.html

**Review by sampan501 — APPROVED** [review]
